### PR TITLE
Memoization function

### PR DESCRIPTION
**Pull request by rudihorn**
This is referencing the computation model specified in #1280. The goal has been to implement a memoization function which takes a computation of the form `'a -> 'b Fiber.t` and returns a computation `'a -> 'b Fiber.t` which additionally performs:
* Dependency tracking
* Output caching with early cut-off
* Cycle detection

Currently the memoization function has been implemented and is able to perform these three functions. An example of how to use the memoization functions is shown in `test/unit-tests/memoize.mlt`. Below is an example of how one could use the system to implement a Fibonacci sequence which is automatically cached between computations.

```
let mfib = 
  let mfib = CRef.deferred () in
  let compfib x =
    let mfib = CRef.get mfib in
    counter := !counter + 1;
    if x <= 1 then
      Fiber.return x
    else
      mfib (x - 1)
      >>= (fun r1 ->
        mfib (x - 2)
        >>| fun r2 -> r1 + r2) in
  memoization iimemcached "fib" int_input_spec int_output_spec compfib |> CRef.set mfib;
  CRef.get mfib;;
```

There has been a little bit of progress to change `build_system.ml` so that the actual build process is a little more abstracted to make it easier to drop in the `memoization` function and remove the current caching etc. This bit is still a little bit of a mess however, and in essence it just breaks existing caching (for now). 

Feedback welcome.

### CODE REVIEW

File: src/memoization.ml [hunk 1]
```diff
@@ -0,0 +1,311 @@
+open !Stdune
+open Import
+(* open Staged *)
+open Fiber.O
+
+type name = string
+type ser_input = string
```
Comment (rudihorn) [inline src/memoization.ml]: Is it worth changing the serialization format to s expressions before we integrate things? I think that should make it easier to have input specs that handle pairs / lists of inputs etc.

Comment (diml) [reply src/memoization.ml]: In practice, we will probably just use marshal as it's faster.

File: src/memoization.ml [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+  (* set up the fiber so that it both has an up to date call stack
+     as well as an empty dependency table *)
+  let wrap_fiber (add_stack : name * ser_input) id (f : 'a Fiber.t) =
+    (* transform f so it returns the dependencies after the computation *)
+    let f = f >>= fun res -> Fiber.Var.get_exn dep_key >>= (fun deps -> Fiber.return (res,!deps)) in
+    (* change f so it sets up a new dependency table before computation *)
+    let f = Fiber.Var.set dep_key (ref []) f in
+    (* set the context so that f has the call stack  *)
+    get_call_stack_int
+      >>| (fun (stack, set) -> add_stack :: stack, Id.Set.add id set) (* add top entry *)
```
Comment (diml) [inline src/memoization.ml]: Let's use a record rather than a pair, to make things more explicit

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/memoization.ml [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+  default_compare = Output;
+}
+let rfn _ =  "<fun>"
+let eager_function_output_spec = {
+  serialize = rfn;
+  print = rfn;
+  default_compare = Recompute;
+}
+
+type update_cache = unit -> ser_output Fiber.t
```
Comment (diml) [inline src/memoization.ml]: I'm not really following what this is

File: src/memoization.ml [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+let int_input_spec = {
+  serialize = string_of_int;
+  print = string_of_int;
+  not_equal = fun i1 i2 -> i1 <> i2;
+}
+
+
+
+type compare_policy =
+  | Output
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.ml]: I have changed this, but I don't use `'a` yet, as I think this may be difficult to do as the policy is saved in the global cache.

File: src/memoization.ml [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+      >>| Option.value ~default:([], Id.Set.empty) (* default call stack is empty *)
+
+  let get_call_stack =
+    get_call_stack_int >>| (fun (stack,_) -> stack)
+
+  (* set up the fiber so that it both has an up to date call stack
+     as well as an empty dependency table *)
+  let wrap_fiber (add_stack : name * ser_input) id (f : 'a Fiber.t) =
+    (* transform f so it returns the dependencies after the computation *)
+    let f = f >>= fun res -> Fiber.Var.get_exn dep_key >>= (fun deps -> Fiber.return (res,!deps)) in
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.ml]: Changed

File: src/memoization.ml [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+}
+
+let global_cache_table : (name * ser_input, global_cache_info) Hashtbl.t = Hashtbl.create 256
+
+module Id = struct
+  include IdGen
+
+  let idgen = IdGen.create ()
+  let gen () = IdGen.gen idgen 
+end
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.ml]: Changed this

**Review by ghost — CHANGES_REQUESTED** [review]

File: runtestmem.sh [hunk 7]
```diff
@@ -0,0 +1,2 @@
+export OCAMLRUNPARAM='l=80000000'
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply runtestmem.sh]: Oh right this file was only for making it easier for me to test. Because the tests are run with `make test` there is no need for this to be pushed.

File: src/build_system.ml [hunk 8]
```diff
@@ -1319,316 +1178,470 @@ let update_universe t =
...
+  let on_error exn =
+    Dep_path.reraise exn (Path path) in
+  (fun () ->
+    get_file_spec t ~loc:loc path
+    >>= (function
+        | None -> (* file already exists *) Fiber.return ()
+        | Some file -> (* build *) build_file_spec file >>| ignore)
+  ) |> Fiber.with_error_handler ~on_error
+
+let _prepare_file t path =
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/build_system.ml]: This code may be nice to register as computation at a later stage, even if it isn't used currently in the code. It essentially builds a files dependencies without building the file itself.

File: src/build_system.ml [hunk 9]
```diff
@@ -448,6 +385,12 @@ type t =
   ; hook : hook -> unit
   ; (* Package files are part of *)
     packages : Package.Name.t Path.Table.t
+  ; load_dir_cache : Path.Set.t Memoize.t
+  (* memoized functions *)
+  ; prepare_rule_def : (Internal_rule.t * Static_deps.t * (unit, Action.t) Build.t, Action.t * Deps.t) CRef.comp
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/build_system.ml]: I've now fixed this, but it does mean that rule.static_deps (which is defined as lazy) is evaluated twice, once in `prepare_rule` and once in `build_rule`. Otherwise one would need to memoize this.

File: src/build_system.ml [hunk 10]
```diff
@@ -1319,316 +1178,470 @@ let update_universe t =
...
+  Deps.paths deps |> parallel_iter_path_set ~f 
+
+let prepare_rule t (rule, static_deps, build : Internal_rule.t * Static_deps.t * ('a, 'b) Build.t) : ('b * Deps.t) Fiber.t =
+  (* get the static dependencies needed before we can call build exec*)
+  let rule_deps = Static_deps.rule_deps static_deps in
+
+  rule.start_rule (); (* legacy for hooks *)
+  (* first compute rule dependencies*)
+  parallel_iter_deps rule_deps ~f:(fun f -> CRef.get t.build_file_def (f,rule.loc))
+  (* then execute the build *)
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/build_system.ml]: Fixed

File: src/build_system.ml [hunk 11]
```diff
@@ -116,7 +86,8 @@ module Internal_rule = struct
     ; mode             : Dune_file.Rule.Mode.t
     ; loc              : Loc.t option
     ; dir              : Path.t
-    ; mutable exec     : Exec_status.t
+    ; start_rule       : unit -> unit (* legacy for hooks *)
+    ; run_rule         : Action.t -> Deps.t -> unit Fiber.t (* the actual computation *)
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/build_system.ml]: I've taken `start_rule` out, but taking out `run_rule` is a little more tricky as it depends on `env` and the root rule doesn't have a valid `env`. Should we do this using an option type for the environment or should we just leave it?

File: src/memoization.ml (hunk 3 above)
Comment (rudihorn) [reply src/memoization.ml]: Added a comment for this.

File: src/memoization.ml [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+(* open Staged *)
+open Fiber.O
+
+type name = string
+type ser_input = string
+type ser_output = string
+
+type 'a input_spec = {
+  serialize : 'a -> ser_input;
+  print : 'a -> string;
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.ml]: currently I'm only using `serialize`, but this should probably be changed to also include the "human readable" form. I'm wondering if it is better to store this as a lazy function (which would prevent the input from being garbage collected) or as the printed input (which would require computation of `print` for every function).

File: src/build_system.ml (hunk 11 above)
[the remark this answers is not in the document]
Comment (diml) [reply src/build_system.ml]: We can just use `Env.empty` for the root rule

File: src/build_system.ml (hunk 8 above)
[the remark this answers is not in the document]
Comment (diml) [reply src/build_system.ml]: Ok. Could you add a comment? Otherwise, future readers will wonder why it's here.

File: src/build_system.ml (hunk 9 above)
[the remark this answers is not in the document]
Comment (diml) [reply src/build_system.ml]: Hmm, evaluating it twice doesn't seem good. We should keep it separate then

File: src/build_system.ml (hunk 11 above)
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/build_system.ml]: I saw it assumed an `Env.t option` in any case, so I have now done this.

File: src/build_system.ml (hunk 9 above)
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/build_system.ml]: I've decided to introduce a `Once` module, which similarly to lazy only evaluates something when `Once.get` is called, but that also caches the result so `Once.get` only causes the expression to be evaluated a single time.

File: src/memoization.mli [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+  (** Get the inputs of all entries with the specified name. **)
+  val filter : name:name -> memoized_cycle_exception -> ser_input list
+end
+
+module Memoize : sig
+  type 'a t
+
+  (** Create a cache used by the memoization for the specific output type.
+      Each output type requires its own cache. *)
+  val create_cache : unit -> 'a t
```
Comment (diml) [inline src/memoization.mli]: It seems to me that we could either:

- create one cache per memoized function
- make the `output_spec` type opaque and store the cache in it

Either would provide a nicer API as the user wouldn't have to manually create the cache.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/stdune/id.mli [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+type t
+
+(** Convert an ID to an integer. *)
+val to_int : t -> int
+
+(** The internal IdGen module which generates an
+    id using an instance of an ID generator for
+    the next ID. *)
+module IdGen : sig
```
Comment (diml) [inline src/stdune/id.mli]: I don't think this should be exposed, it's not used anywhere.

File: src/stdune/id.mli [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+end
+
+module Set : Set_intf.S with type elt = t
+
+module type IdMod = sig
+  type t
+
+  module Set : Set_intf.S with type elt = t
+
+  val idgen : IdGen.idgen
```
Comment (diml) [inline src/stdune/id.mli]: This shouldn't be exposed either.

File: src/stdune/id.mli [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+  val peek : idgen -> t
+  val create : unit -> idgen
+  val to_int : t -> int
+  val gen : idgen -> t
+end
+
+module Set : Set_intf.S with type elt = t
+
+module type IdMod = sig
```
Comment (diml) [inline src/stdune/id.mli]: The usual convention is to call this `S`.

File: src/stdune/id.ml [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+  val peek : unit -> t
+  val to_int : t -> int
+  val compare : t -> t -> Ordering.t
+end
+
+module Make () : IdMod = struct
+  module Set = Set.Make(M)
+  type t = int
+
+  let idgen = IdGen.create ()
```
Comment (diml) [inline src/stdune/id.ml]: Why not do this directly? i.e.:

```ocaml
let next = ref  0
let gen () = ...
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/stdune/id.ml [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+  val idgen : IdGen.idgen
+  val gen : unit -> t
+  val peek : unit -> t
+  val to_int : t -> int
+  val compare : t -> t -> Ordering.t
+end
+
+module Make () : IdMod = struct
+  module Set = Set.Make(M)
```
Comment (diml) [inline src/stdune/id.ml]:
```suggest
  module Set = Int.Set
```

File: src/stdune/id.mli (hunk 16 above)
Comment (rudihorn) [reply src/stdune/id.mli]: updated

File: src/stdune/id.mli (hunk 14 above)
Comment (rudihorn) [reply src/stdune/id.mli]: This isn't used, but I thought it might be useful if some object would like none local ids, e.g. if internal rules would like to generate local ids for dependencies it could do something like:

```
type t = {
  idgen : IdGen.t
}

...
let add_dep rule =
  ... let id = IdGen.gen rule.idgen in ...
```

Otherwise I can simplify it as you mentioned below, getting rid of all the id gen things.

File: src/stdune/id.ml (hunk 18 above)
Comment (rudihorn) [reply src/stdune/id.ml]: Changed

File: src/memoization.mli (hunk 13 above)
Comment (rudihorn) [reply src/memoization.mli]: The new behaviour is to create the cache in the memoization function.

File: src/memoization.mli [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+type dep_node
+
+(** A stack frame within a computation. *)
+type stack_frame = {
+  name : name;
+  input : ser_input;
+  dep_node : dep_node;
+}
+
+module CycleException : sig
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.mli]: Changed.

File: src/memoization.mli [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+      @raise Exn.Fatal_error if [set] was already called *)
+  val set : ('a, 'b) comp -> ('a, 'b) t -> unit
+
+  (** [get comp] returns the [f] if [set comp f] was called.
+      @raise Exn.Fatal_error if [set] has not been called yet. *)
+  val get : ('a, 'b) comp -> ('a, 'b) t
+end
+
+type dep_info
+type dep_node
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.mli]: Removed.

File: src/memoization.mli [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+  | No_cutoff (** The output should not be compared and the function should always be recomputed. *)
+  | Cutoff (** The outputs should be compared and only recomputed if they differ. *)
+
+type 'a output_spec = {
+  serialize : 'a -> ser_output;
+  print : 'a -> string;
+  cutoff_policy : cutoff_policy;
+}
+
+(** Computation References are used to allow the recursive definitions of files. Similar to IVar's,
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.mli]: Changed.

File: src/stdune/id.mli (hunk 14 above)
Comment (diml) [reply src/stdune/id.mli]: Yh, let's simplify it. I prefer to keep dune lean for now.

File: src/stdune/id.mli [hunk 22]
```diff
@@ -0,0 +1,23 @@
+open !Set
```
Comment (diml) [inline src/stdune/id.mli]: This can be deleted

File: src/stdune/id.mli [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+
+module type S = sig
+  type t
+
+  module Set : Set_intf.S with type elt = t
```
Comment (diml) [inline src/stdune/id.mli]:
```suggestion
  module Set : Set.S with type elt = t
```

(`X_intf` modules are for internal purpose only, they are just to avoid duplicating signatures. In general they shouldn't be used outside of `X`)

File: src/stdune/once.mli [hunk 24]
```diff
@@ -0,0 +1,27 @@
+(** Cached lazy evaluation *)
```
Comment (diml) [inline src/stdune/once.mli]: This module is now unused, so we can remove it from this PR.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/stdune/once.mli (hunk 24 above)
Comment (rudihorn) [reply src/stdune/once.mli]: removed it.

File: src/stdune/id.mli (hunk 23 above)
Comment (rudihorn) [reply src/stdune/id.mli]: Changed.

File: src/stdune/id.mli (hunk 22 above)
Comment (rudihorn) [reply src/stdune/id.mli]: Removed.

**Comment by rgrinberg** [comment]
@rudihorn Once this is ready to be merged, could you remove the WIP tag? I'd like to give this a read when it's ready. To make sure I understand the basics.

File: src/memoization.ml [hunk 25]
```diff
@@ -0,0 +1,376 @@
...
+    ex.stack
+    |> List.map ~f:(fun v -> Stack_frame.di v)
+    |> List.map ~f:(fun (f : dep_info) -> Format.sprintf "%s %s" f.name f.input)
+end
+
+
+let global_dep_dag = Dag.create ()
+let global_dep_table : (FnId.t * ser_input, dep_node) Hashtbl.t = Hashtbl.create 256
+
+let global_id_table : (name, FnId.t) Hashtbl.t = Hashtbl.create 256
```
Comment (diml) [inline src/memoization.ml]: Instead of FnId, you can use something like `module Function_name = Interned.Make(...)`

File: src/stdune/list.ml [hunk 26]
```diff
@@ -87,6 +87,9 @@ let rec find l ~f =
   | [] -> None
   | x :: l -> if f x then Some x else find l ~f
 
+let exists l ~f =
+  find l ~f |> Option.is_some
```
Comment (rgrinberg) [inline src/stdune/list.ml]: This function is already defined. `ListLabels.exists` or `List.exists`.

**Review by rgrinberg — APPROVED** [review]

File: src/memoization.ml (hunk 25 above)
Comment (rudihorn) [reply src/memoization.ml]: Changed.

File: src/memoization.ml [hunk 27]
```diff
@@ -0,0 +1,376 @@
+open !Stdune
+open Import
+(* open Staged *)
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.ml]: Removed.

File: src/memoization.ml [hunk 28]
```diff
@@ -0,0 +1,376 @@
...
+
+  let filter ~(name:name) ex =
+    ex.cycle
+    |> List.filter ~f:(fun (f : dep_node) -> (Dag.get f).name = name)
+    |> List.map ~f:(fun (f : dep_node) -> (Dag.get f).input)
+
+  let serialize ex =
+    ex.cycle
+    |> List.map ~f:Dag.get
+    |> List.map ~f:(fun (f : dep_info) -> Format.sprintf "%s %s" f.name f.input)
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.ml]: Changed.

File: src/memoization.ml [hunk 29]
```diff
@@ -0,0 +1,376 @@
...
+    | Some id -> id
+
+  let get_deps (name : name) (inp : ser_input) =
+    let id = get_name_id name in
+    let dep_info = get_dependency_node id name inp in
+    Option.map ~f:(fun r -> r.last_deps |> List.map ~f:(fun (n,_u) ->
+      let node = Dag.get n in
+      node.name,node.input)) (Dag.get dep_info).cache
+
+  let rec list_any l =
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.ml]: Changed.

File: src/memoization.ml [hunk 30]
```diff
@@ -0,0 +1,376 @@
...
+
+  let get_run_id =
+    Fiber.Var.get_exn run_id
+
+  (* set up the fiber so that it both has an up to date call stack
+     as well as an empty dependency table *)
+  let wrap_fiber (stack_frame : Stack_frame.t) id (f : 'a Fiber.t) =
+    (* set the context so that f has the call stack  *)
+    get_call_stack_int
+      >>| (fun (stack, set) -> stack_frame :: stack, Id.Set.add set id) (* add top entry *)
```
[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization.ml]: Changed.

File: src/memoization_cached.ml [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+
+type 'a t = Id.t * 'a Fiber.t -> 'a Fiber.t
+
+let id_input_spec =
+  let ser x = Id.to_int x |> string_of_int in
+  let ne x y = x <> y in
+  {
+    serialize = ser;
+    print = ser;
+    not_equal = ne;
```
[the remark this answers is not in the document]
Comment (emillon) [reply src/memoization_cached.ml]: yes, more precisely `let serialize = ... in let not_equal = ... in { serialize; print = serialize; not_equal}`. Otherwise I'm fine with the current version.

[the remark this answers is not in the document]
Comment (rudihorn) [reply src/memoization_cached.ml]: I applied this in a few places where it seemed to make sense

File: src/build_system.ml [hunk 32]
```diff
@@ -1521,35 +1527,33 @@ let package_deps t pkg files =
...
+        )
       end
   in
   let open Build.O in
   Build.paths_for_rule files >>^ fun () ->
   (* We know that at this point of execution, all the relevant ivars
      have been filled *)
-  Path.Set.fold files ~init:Package.Name.Set.empty ~f:loop_deps
+  Path.Set.fold files ~init:Package.Name.Set.empty
+    ~f:(fun f acc -> loop_deps f acc |> Memoize.run_memoize |> Fiber.run)
```
Comment (diml) [inline src/build_system.ml]: Just noticed this, we need to get rid of this `Fiber.run`

Comment (diml) [reply src/build_system.ml]: It's now gone

**Comment by diml** [comment]
I did a bit more work on this PR and I now get very similar timings between the base and tip of this PR. For null builds of the platform, there is now only a 0.1s slowdown, which is likely due to the new cycle detection algorithm. This is fine given that the one in master is broken. We can now rebase and merge this PR.

**Review by ghost — APPROVED** [review]

**Comment by diml** [comment]
@rudihorn there is an issue with the signature, could you update the email in `author` field of the commit and force-push?

@rgrinberg @emillon could you have another look as I changed a couple of things (and @rudihorn as well if you want).

File: src/dag/dag.ml [hunk 33]
```diff
@@ -0,0 +1,152 @@
...
+    let m = dag.arcs in
+    let delta = min (float m ** (1.0 /. 2.0)) (float n ** (2.0 /. 3.0)) |> int_of_float in
+    delta
+
+  let add dag v w =
+    let delta = delta dag in
+
+    dag.arcs <- dag.arcs + 1;
+
+    (* Printf.printf "new add %s -> %s\n%!" v.input w.input; *)
```
Comment (emillon) [inline src/dag/dag.ml]:
```suggestion
```

File: src/dag/dag.ml [hunk 34]
```diff
@@ -0,0 +1,152 @@
...
+      level = 1;
+      deps = [];
+      rev_deps = [];
+    }
+
+  let delta dag =
+    let n = dag.num in
+    let m = dag.arcs in
+    let delta = min (float m ** (1.0 /. 2.0)) (float n ** (2.0 /. 3.0)) |> int_of_float in
+    delta
```
Comment (emillon) [inline src/dag/dag.ml]: you can inline the definition here.

File: src/dag/dag.ml [hunk 35]
```diff
@@ -0,0 +1,152 @@
...
+    and btraverse x _y acc =
+      if x.info.id = w.info.id then raise (Cycle acc);
+      arcs := !arcs + 1;
+      if !arcs >= delta then begin
+        w.info.level <- v.info.level + 1;
+        w.info.rev_deps <- [];
+        b := [];
+        true
+      end else begin
+        if Int.Set.mem !marked_ids x.info.id |> not then
```
Comment (emillon) [inline src/dag/dag.ml]: the `|> not` is a bit hard to see, I'd suggest to invert the branches. (this can be rewritten as `mem ... && bvisit ...` but since it is effectful, that does not seem to be a good idea)

File: src/dag/dag.ml [hunk 36]
```diff
@@ -0,0 +1,152 @@
...
+
+    (* step 5 *)
+    v.info.deps <- w :: v.info.deps;
+    if v.info.level = w.info.level then
+      w.info.rev_deps <- v :: w.info.rev_deps
+
+  let children node = node.info.deps
+
+  let pp_list pp_content fmt v =
+    List.iter ~f:(fun f -> Format.fprintf fmt "%a;@, " pp_content f) v
```
Comment (emillon) [inline src/dag/dag.ml]: Would it make sense to use `Pp.list` here?

File: src/dag/dag_intf.ml [hunk 37]
```diff
@@ -0,0 +1,52 @@
...
+
+  (** Type of values attached to nodes. *)
+  type value
+
+  type node =
+    { data : value
+    ; info : node_info
+    }
+
+  (** An cycle has been found while adding an arc. *)
```
Comment (emillon) [inline src/dag/dag_intf.ml]:
```suggestion
  (** A cycle has been found while adding an arc. *)
```

File: src/dag/dag.ml (hunk 36 above)
Comment (emillon) [reply src/dag/dag.ml]: (or Fmt.list)

File: src/memo/memo.ml [hunk 38]
```diff
@@ -0,0 +1,335 @@
...
+          Fiber.Ivar.read fut
+        else
+          recompute t inp dep_node
+      | Done cv ->
+        Cached_value.get cv >>= function
+        | Some v -> Fiber.return v
+        | None -> recompute t inp dep_node
+
+  let peek t inp =
+    (* XXX: this doesn't add a reverse dependency, which is wrong *)
```
Comment (emillon) [inline src/memo/memo.ml]: Should this be addressed? I'd rather replace this with a pointer to a bug or something a bit actionable. Or just remove the XXX.

File: src/memo/memo.mli [hunk 39]
```diff
@@ -0,0 +1,47 @@
...
+  val name : t -> string
+  val input : t -> Sexp.t
+end
+
+module Cycle_error : sig
+  type t
+
+  exception E of t
+
+  (** Get the list of stack frame in this cycle. *)
```
Comment (emillon) [inline src/memo/memo.mli]:
```suggestion
  (** Get the list of stack frames in this cycle. *)
```

File: src/memo/memo.mli [hunk 40]
```diff
@@ -0,0 +1,47 @@
...
+  val get : t -> Stack_frame.t list
+
+  (** Return the stack leading to the node which raised the cycle. *)
+  val stack : t -> Stack_frame.t list
+end
+
+(** Restart the system. Cached values with a [Current_run] lifetime
+    are forgotten, pending computation are cancelled. *)
```
Comment (emillon) [inline src/memo/memo.mli]:
```suggestion
    are forgotten, pending computations are cancelled. *)
```

File: src/memo/memo_intf.ml [hunk 41]
```diff
@@ -0,0 +1,52 @@
...
+  (** [create name ?lifetime ouput_spec f] creates a memoized version
+      of [f]. The result of [f] for a given input is cached, so that
+      the second time [exec t x] is called, the previous result is
+      re-used if possible.
+
+      [exec t x] tracks what calls to other memoized function [f x]
+      performs. When the result of such dependent call changes, [exec t
+      x] will automatically recomputes [f x].
+
+      Running the computation may raise [MemCycle] if a cycle is
```
Comment (emillon) [inline src/memo/memo_intf.ml]: Is this `Memo.Cycle_Error`?

File: src/static_deps.mli [hunk 42]
```diff
@@ -2,6 +2,10 @@ open! Import
 
 (** A simple wrapper around [Deps.t], where some dependencies are recorded as
     "rule deps" and other as "action deps". *)
+
+(* Action Dependencies: Files the external commands are expected to access *)
+(* Rule Dependencies: Files needed in order to compute the action to execute and its dependencies *)
+
```
Comment (emillon) [inline src/static_deps.mli]: Could this be added to the above docstring?

File: src/stdune/fdecl.ml [hunk 43]
```diff
@@ -0,0 +1,17 @@
+type 'a state =
+  | Unset
+  | Set of 'a
+
+type 'a t = { mutable state : 'a state }
```
Comment (emillon) [inline src/stdune/fdecl.ml]: What do you think about adding a label that can be displayed in error messages?

**Review by emillon — COMMENTED** [review]
Thanks! I have a couple minor comments but I look forward to having this merged :+1:

File: src/dag/dag.ml (hunk 33 above)
Comment (diml) [reply src/dag/dag.ml]: removed

File: src/dag/dag.ml (hunk 34 above)
Comment (diml) [reply src/dag/dag.ml]: indeed, done

File: src/dag/dag.ml (hunk 36 above)
Comment (diml) [reply src/dag/dag.ml]: Indeed, I updated the code

File: src/dag/dag.ml (hunk 35 above)
Comment (diml) [reply src/dag/dag.ml]: Agreed, I inverted the branches

File: src/dag/dag_intf.ml (hunk 37 above)
Comment (diml) [reply src/dag/dag_intf.ml]: fixed

File: src/top_closure.ml [hunk 44]
```diff
@@ -44,6 +52,38 @@ module Make(Keys : Keys) = struct
     match iter_elts elements ~temporarily_marked:Keys.empty with
     | Ok () -> Ok (List.rev !res)
     | Error elts -> Error elts
+
+  let top_closure_f ~key ~deps elements =
```
Comment (rgrinberg) [inline src/top_closure.ml]: Is it worth it to get rid of this duplicate with a functor over a monad? This is almost exactly the same as `top_closure`.

File: src/memo/memo.ml (hunk 38 above)
Comment (diml) [reply src/memo/memo.ml]: Indeed, I opened a ticket and added a reference

File: src/memo/memo.mli (hunk 39 above)
Comment (diml) [reply src/memo/memo.mli]: fixed

File: src/memo/memo_intf.ml (hunk 41 above)
Comment (diml) [reply src/memo/memo_intf.ml]: indeed

File: src/memo/memo.mli (hunk 40 above)
Comment (diml) [reply src/memo/memo.mli]: fixed

File: src/static_deps.mli (hunk 42 above)
Comment (diml) [reply src/static_deps.mli]: Sure, done

File: src/stdune/fdecl.ml (hunk 43 above)
Comment (diml) [reply src/stdune/fdecl.ml]: I suppose the backtrace should be enough

File: src/top_closure.ml (hunk 44 above)
Comment (diml) [reply src/top_closure.ml]: Yh, that make sense

File: src/build_system.ml [hunk 45]
```diff
@@ -1528,29 +1552,34 @@ let package_deps t pkg files =
...
+              ~init:acc ~f:loop
       end
   in
   let open Build.O in
   Build.paths_for_rule files >>^ fun () ->
   (* We know that at this point of execution, all the relevant ivars
      have been filled *)
-  Path.Set.fold files ~init:Package.Name.Set.empty ~f:loop_deps
+  Path.Set.fold files ~init:Package.Name.Set.empty
+    ~f:(fun f acc -> loop_deps f acc)
```
Comment (rgrinberg) [inline src/build_system.ml]: Curious why this change was made. The eta reduced version seemed preferable to me

File: src/build_system.ml [hunk 46]
```diff
@@ -1528,29 +1552,34 @@ let package_deps t pkg files =
...
-            (Lazy.force ir.static_deps)
-        in
-        Path.Set.fold
-          (Deps.path_union action_deps dyn_deps)
-          ~init:acc ~f:loop
+        match Fiber.Once.peek ir.static_deps with
+        | None ->
+          (* We know that at this point of execution, all the relevant
+             ivars have been filled *)
+          assert false
```
Comment (rgrinberg) [inline src/build_system.ml]: Having `Fiber.Once.peek_exn` would make this code a little cleaner I think - a little less nesting to the right.

File: src/stdune/fdecl.ml (hunk 43 above)
Comment (emillon) [reply src/stdune/fdecl.ml]: Sure!

File: src/top_closure.ml (hunk 44 above)
Comment (diml) [reply src/top_closure.ml]: done

File: src/build_system.ml (hunk 45 above)
Comment (diml) [reply src/build_system.ml]: Oops, it's just a left-over. Some code was added and then removed, but it wasn't simplified afterwards.

File: src/build_system.ml (hunk 46 above)
Comment (diml) [reply src/build_system.ml]: Indeed, done

**Comment by rgrinberg** [comment]
@diml feel free to merge after fixing bootstrap. I'd still like to study some parts of this PR, but I don't think we should delay merging. Rebasing these pr's over and over is a waste of time.
